lib/ohai/plugins/passwd.rb

```diff
@@ -42,6 +42,72 @@ def fix_encoding(str)
   end
 
   collect_data(:windows) do
-    # Etc returns nil on Windows
+    require "wmi-lite/wmi" unless defined?(WmiLite::Wmi)
+
+    unless etc
+      etc Mash.new
+
+      wmi = WmiLite::Wmi.new
+
+      etc[:passwd] = Mash.new
+      users = wmi.query("SELECT * FROM Win32_UserAccount Where LocalAccount = True")
+      users.each do |user|
+        uname = user["Name"].strip.downcase
+        Ohai::Log.debug("processing user #{uname}")
+        etc[:passwd][uname] = Mash.new
+        wmi_obj = user.wmi_ole_object
+        wmi_obj.properties_.each do |key|
+          etc[:passwd][uname][key.name.downcase] = user[key.name]
+        end
+      end
+
+      etc[:group] = Mash.new
+      groups = wmi.query("SELECT * FROM Win32_Group Where LocalAccount = True")
+      groups.each do |group|
+        gname = group["Name"].strip.downcase
+        Ohai::Log.debug("processing group #{gname}")
+        etc[:group][gname] = Mash.new
+        wmi_obj = group.wmi_ole_object
+        wmi_obj.properties_.each do |key|
+          etc[:group][gname][key.name.downcase] = group[key.name]
+        end
+
+        # This is the primary reason that we're using WMI instead of powershell
+        # cmdlets - the powershell start up cost is huge, and you *must* do this
+        # query for every. single. group. individually.
+
+        # The query returns nothing unless you specify domain *and* name, it's
+        # a path, not a set of queries.
+        subq = "Win32_Group.Domain='#{group["Domain"]}',Name='#{group["Name"]}'"
+        members = wmi.query(
+          "select * from Win32_GroupUser Where GroupComponent=\"#{subq}\""
+        )
+        etc[:group][gname]["members"] = members.map do |member|
+          mi = {}
+          info = Hash[member["partcomponent"].split(",").map { |x| x.split("=") }]
+          if info.keys.any? { |x| x.match?("Win32_UserAccount") }
+            mi["type"] = :user
+          else
+            # Note: the type here is actually Win32_SystemAccount, because,
+            # that's what groups are in the Windows universe.
+            mi["type"] = :group
+          end
+          # WARNING! HERE BE DRAGONS!
+          # info['name'] will be a quoted string becuase we parsed a string
+          # of foo="bar",baz="crazy\"\ string"
+          # eval here is sketchy, but I can't think of a better way to get a
+          # raw string out of this. But if someone gets the right shit into
+          # a group name, you could make Chef do bad things. But then again,
+          # if you can do that, you're already the admin. But then again
+          # with TargetMode, this is probably extra terrifying. Someone
+          # give me a better way.
```

Comment:
I think `String.undump` could be useful here but it's kind of finicky:

```
irb(main):026:0> "\"hello\"".undump
=> "hello"
irb(main):027:0> "foo".undump
Traceback (most recent call last):
        5: from /Users/pete/.asdf/installs/ruby/2.6.6/bin/irb:23:in `<main>'
        4: from /Users/pete/.asdf/installs/ruby/2.6.6/bin/irb:23:in `load'
        3: from /Users/pete/.asdf/installs/ruby/2.6.6/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):27
        1: from (irb):27:in `undump'
RuntimeError (invalid dumped string; not wrapped with '"' nor '"...".force_encoding("...")' form)
irb(main):028:0> "\"crazy\" string\"".undump
Traceback (most recent call last):
        6: from /Users/pete/.asdf/installs/ruby/2.6.6/bin/irb:23:in `<main>'
        5: from /Users/pete/.asdf/installs/ruby/2.6.6/bin/irb:23:in `load'
        4: from /Users/pete/.asdf/installs/ruby/2.6.6/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        3: from (irb):28
        2: from (irb):28:in `rescue in irb_binding'
        1: from (irb):28:in `undump'
RuntimeError (invalid dumped string; not wrapped with '"' nor '"...".force_encoding("...")' form)
```

Do you have examples of what's actually returned from this?

Reply:
This almost seems like a CSV file with `=` as the field separator and `,` as the line delimiter...

Reply:
```
"partcomponent"=>"\\\\MY-HOSTNAME\\root\\cimv2:Win32_SystemAccount.Domain=\"MY-HOSTNAME\",Name=\"Authenticated Users\""
```
is an example. I replaced the hostname with MY-HOSTNAME (which is always in all-caps here, because Windows, I guess)

Reply:
actually I'm only undumping the *value* here not the key. the value seems less encoded, just surrounding quotes. So undump probably works.

```
irb(main):008:0> "\"Authenticated Users\"".undump
=> "Authenticated Users"
```

will update and test

Reply:
It seems to work for your example:

```
irb(main):042:0> example = "\\\\MY-HOSTNAME\\root\\cimv2:Win32_SystemAccount.Domain=\"MY-HOSTNAME\",Name=\"Authenticated Users\""
=> "\\\\MY-HOSTNAME\\root\\cimv2:Win32_SystemAccount.Domain=\"MY-HOSTNAME\",Name=\"Authenticated Users\""
irb(main):043:0> example.split(",").map { |s| s.split("=") }.map {|a,b| [a, b.undump] }
=> [["\\\\MY-HOSTNAME\\root\\cimv2:Win32_SystemAccount.Domain", "MY-HOSTNAME"], ["Name", "Authenticated Users"]]
```

Reply:
... and in case it gets more complicated:

```
irb(main):047:0> CSV.parse(example, row_sep: ',', col_sep: '=')
=> [["\\\\MY-HOSTNAME\\root\\cimv2:Win32_SystemAccount.Domain", "MY-HOSTNAME"], ["Name", "Authenticated Users"]]
```

Reply:
@phiggins thanks! I feel much better having nuked that `eval`. :)

Reply:
TIL dump/undump

Reply:
@lamont-granquist right? Me too!